macaw-aarch32/src/Data/Macaw/ARM/Semantics/TH.hs

```diff
@@ -594,18 +799,18 @@ armAppEvaluator :: M.Endianness
                 -> Maybe (MacawQ ARM.AArch32 t fs Exp)
 armAppEvaluator endianness interps elt =
     case elt of
-      WB.BaseIte bt _ test t f | isPlaceholderType bt -> return $ do
+      WB.BaseIte bt _ test t f | Just wk <- typeAsWriteKind bt -> return $ do
         -- In this case, the placeholder type indicates that
         -- expression is to be translated as a (wrapped) stateful action
-        -- rather than an actual macaw term. This is therefore translated
-        -- as a Haskell if-then-else statement, rather than
-        -- a Mux.
+        -- rather than an actual macaw term. The mux condition is therefore mapped
+        -- across all of the stateful actions
         testE <- addEltTH endianness interps test
-        tE <- addEltTH endianness interps t
-        fE <- addEltTH endianness interps f
-        case all isEager [testE, tE, fE] of
-          True -> liftQ [| return $ concreteIte $(refEager testE) $(refEager tE) $(refEager fE) |]
-          False -> liftQ [| concreteIte <$> $(refBinding testE) <*> $(refBinding tE) <*> $(refBinding fE) |]
+        inConditionalContext $ do
```

Comment:
Interesting that this came back

Reply:
Yeah, we're now unconditionally executing both branches in the Generator monad and only making the writes themselves conditional. This ends up generating unnecessary memory reads for branches that don't get executed, so we need to wrap those reads up as lazily-evaluated values.